Arbitrary Query Filters

afrasier opened the pull request:
* Added post method to awards/summary/
* You can now specify any valid field as a get or post field parameter
* POST requests can now have any combination of filters via OR and AND. User friendly documentation forthcoming

kaitlin commented on usaspending_api/awards/views.py:
```diff
@@ -33,69 +34,70 @@ def get(self, request, uri=None, piid=None, fain=None, format=None):
+    def post(self, request, format=None):
+        fg = FilterGenerator()
+        body_unicode = request.body.decode('utf-8')
+        body = json.loads(body_unicode)
```
Can you add exception handling for a poorly formed request object?

kaitlin commented on usaspending_api/common/api_request_utils.py:
```diff
@@ -0,0 +1,142 @@
+
+    def create_q_from_filter_list(self, filter_list, combine_method='AND'):
+        q_object = Q()
+        for filt in filter_list:
```
what are your thoughts on how we should sanitize the input for different filters? Perhaps for now we could raise a 400 somewhere in this block and in the next sprint try and surface some detailed info about which filter or parameter value was malformed.

afrasier replied:
Yeah, sanity checks for operators are pretty easy (i.e. for __range your value must have a list of length 2) but as far as a malformed field name that can be a bit trickier. For now I'll implement the value/operator validations and 400 if the fieldname is somehow bad.

afrasier commented:
@kaitlin I've added a validator for post requests that will spit out 400 responses with an error message if the request is malformed or uses an invalid operator, or uses a bad value for certain operators. Added tests for those (though I need to up the robustness of the tests in the future)

If they specify a field that doesn't exist (or something related, like a foreign key lookup) it will spit out the standard django error about it

Let me know what else needs doing, or go ahead and approve+merge :) Thanks! 👍
